File: engines/cpp/src/ir_vm/parser/parser.cpp
#include "Parser.hpp"

namespace parser
{
    void Parser::setSource(const std::string &source)
    {
        this->source = source;
    }

    void Parser::advance()
    {
        this->current++;
    }

    inline instruction::Instruction Parser::emitByte(instruction::OpCode opcode, int argument)
    {
        return instruction::Instruction{opcode, argument};
    };

    std::vector<instruction::Instruction> Parser::parse()
    {
        this->instruction.clear();
        while (!loop_stack.empty())
            loop_stack.pop();
        this->current = 0;

        while (current < this->source.size())
        {
            char c = source.at(current);
            switch (c)
            {
            case '+':
            {
                this->instruction.push_back(this->emitByte(instruction::OpCode::ADD, 1));
                break;
            }
            case '-':
            {
                this->instruction.push_back(this->emitByte(instruction::OpCode::ADD, -1));
                break;
            }
            case '>':
            {
                this->instruction.push_back(this->emitByte(instruction::OpCode::MOVE, 1));
                break;
            }
            case '<':
            {
                this->instruction.push_back(this->emitByte(instruction::OpCode::MOVE, -1));
                break;
            }
            case ',':
            {
                this->instruction.push_back(this->emitByte(instruction::OpCode::INPUT, 0));
                break;
            }
            case '.':
            {
                this->instruction.push_back(this->emitByte(instruction::OpCode::OUTPUT, 0));
                break;
            }
            case '[':
            {
                this->instruction.push_back(this->emitByte(instruction::OpCode::JZ, -1));
                this->loop_stack.push(instruction.size() - 1);
                break;
            }
            case ']':
            {
                if (this->loop_stack.empty())
                {
                    throw ParseError("Unmatched ']' at source position " + std::to_string(current));
                }

                int start_idx = this->loop_stack.top();
                this->loop_stack.pop();
                this->instruction.push_back(this->emitByte(instruction::OpCode::JNZ, start_idx));
                this->instruction[start_idx].argument = instruction.size() - 1;
                break;
            }

            default:
                break;
            }
            advance();
        }
        return this->instruction;
    }

} // namespace parser

File: engines/zig/src/ir_vm/parser/parser.zig
const std: type = @import("std");
const instruction: type = @import("../instructions/opcode.zig");

const errors = @import("../../errors/error.zig");

pub const Parser = struct {
    current: usize,
    source: []const u8,
    array: std.ArrayList(instruction.Instructions),
    loop_stack: std.ArrayList(usize),

    allocator: std.mem.Allocator,

    pub fn init(self: *Parser, source: []const u8, allocator: std.mem.Allocator) !void {
        // std.debug.print("Initializing the parser...\n", .{});
        self.source = source;
        self.allocator = allocator;
        self.current = 0;
        self.array = try std.ArrayList(instruction.Instructions).initCapacity(self.allocator, 1024);
        self.loop_stack = try std.ArrayList(usize).initCapacity(self.allocator, 256);
    }

    pub fn deinit(self: *Parser) void {
        // std.debug.print("Deinitializing the parser...\n", .{});
        self.array.deinit(self.allocator);
        self.loop_stack.deinit(self.allocator);
    }

    pub fn setSource(self: *Parser, source: []const u8) void {
        self.source = source;
        self.current = 0;
    }

    pub fn parse(self: *Parser) errors.Errors![]instruction.Instructions {
        // std.debug.print("Parsing the raw code to IR...\n", .{});
        while (self.current < self.source.len) {
            switch (self.source[self.current]) {
                '+' => {
                    self.array.append(self.allocator, emitByte(instruction.OpCode.ADD, 1)) catch return errors.Errors.ParseError;
                },
                '-' => {
                    self.array.append(self.allocator, emitByte(instruction.OpCode.ADD, -1)) catch return errors.Errors.ParseError;
                },
                '>' => {
                    self.array.append(self.allocator, emitByte(instruction.OpCode.MOVE, 1)) catch return errors.Errors.ParseError;
                },
                '<' => {
                    self.array.append(self.allocator, emitByte(instruction.OpCode.MOVE, -1)) catch return errors.Errors.ParseError;
                },
                '.' => {
                    self.array.append(self.allocator, emitByte(instruction.OpCode.OUTPUT, 0)) catch return errors.Errors.ParseError;
                },
                ',' => {
                    self.array.append(self.allocator, emitByte(instruction.OpCode.INPUT, 0)) catch return errors.Errors.ParseError;
                },
                '[' => {
                    self.loop_stack.append(self.allocator, self.array.items.len) catch return errors.Errors.ParseError;
                    self.array.append(self.allocator, emitByte(instruction.OpCode.JZ, 0)) catch return errors.Errors.ParseError;
                },
                ']' => {
                    const start = self.loop_stack.pop() orelse return errors.Errors.ParseError;
                    self.array.items[start].argument = @intCast(self.array.items.len);
                    self.array.append(self.allocator, emitByte(instruction.OpCode.JNZ, @intCast(start))) catch return errors.Errors.ParseError;
                },
                else => {},
            }
            self.advance();
        }

        return self.array.items;
    }

    fn emitByte(opcode: instruction.OpCode, argument: i32) instruction.Instructions {
        return instruction.Instructions{ .opcode = opcode, .argument = argument };
    }

    fn advance(self: *Parser) void {
        self.current += 1;
    }
};
